File: HR-js/ipk/sortArrMinSwap.js
// https://www.hackerrank.com/challenges/minimum-swaps-2/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays
function main(input) {
    let ip_arr = input.split("\n");
    let len = Number(ip_arr[0]);
    let unsorted_arr = ip_arr[1].split(' ').map(Number);
    let min_swap = findMinSwap(len, unsorted_arr);
    console.log(min_swap);
}

function findMinSwap(len, unsorted_arr) {
    let sorted_arr = unsorted_arr.slice().sort((a, b) => a - b);
    let position_map = new Map();
    sorted_arr.forEach((ele, index) => {
        position_map.set(ele, index);
    });
    let visited = {};
    let total_swap = 0;
    for (let i = 0; i < len; i++) {
        if (visited[i] || position_map[unsorted_arr[i]] === i)
            continue;
        let circle_size = 0;
        let j = i;
        while (!visited[j]) {
            visited[j] = true;
            j = position_map.get(unsorted_arr[j]);
            circle_size++;
        }
        total_swap += (circle_size - 1);
    }
    return total_swap;
}
// main("4\n4 3 1 2");
main("5\n2 3 4 1 5");

File: HR-python/sortArrMinSwap.py
# // https://www.hackerrank.com/challenges/minimum-swaps-2/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays

def findMinSwap(size, unsorted_arr):
    sorted_arr = sorted(unsorted_arr)
    sorted_index_dict = dict()

    for i in range (0, size):
        sorted_index_dict[sorted_arr[i]] = i

    visited_dict = [False]*size
    min_swap = 0
    for i in range (0, size):
        if (visited_dict[i] == True or sorted_index_dict[unsorted_arr[i]] == i):
            continue
        j = i
        circle_size = 0
        while not visited_dict[j]:
            visited_dict[j] = True
            j = sorted_index_dict[unsorted_arr[j]]
            circle_size += 1
        min_swap += (circle_size - 1)
    return min_swap

def main():
    size = 5
    unsorted_arr = [2, 3, 4, 1, 5]
    min_swap = findMinSwap(size, unsorted_arr)
    print(min_swap)

main()



    